netty/src/main/java/io/grpc/netty/NettyClientStream.java

```diff
@@ -183,17 +183,31 @@ private void writeFrameInternal(
         // Add the bytes to outbound flow control.
         onSendingBytes(numBytes);
         writeQueue.enqueue(new SendGrpcFrameCommand(transportState(), bytebuf, endOfStream), flush)
-            .addListener(new ChannelFutureListener() {
-              @Override
-              public void operationComplete(ChannelFuture future) throws Exception {
-                // If the future succeeds when http2stream is null, the stream has been cancelled
-                // before it began and Netty is purging pending writes from the flow-controller.
-                if (future.isSuccess() && transportState().http2Stream() != null) {
-                  // Remove the bytes from outbound flow control, optionally notifying
-                  // the client that they can send more bytes.
-                  transportState().onSentBytes(numBytes);
-                  NettyClientStream.this.getTransportTracer().reportMessageSent(numMessages);
-                }
+            .addListener((ChannelFuture future) -> {
+              // If the future succeeds when http2stream is null, the stream has been cancelled
+              // before it began and Netty is purging pending writes from the flow-controller.
+              if (future.isSuccess() && transportState().http2Stream() == null) {
+                return;
+              }
+
+              if (future.isSuccess()) {
+                // Remove the bytes from outbound flow control, optionally notifying
+                // the client that they can send more bytes.
+                transportState().onSentBytes(numBytes);
+                NettyClientStream.this.getTransportTracer().reportMessageSent(numMessages);
+              } else if (isReady()) {
```

Comment:
As discussed offline, `isReady()` will give false negatives, because of flow control.